File: facefind_front/src/services/alertaService.js
/**
 * Alerta Service - Cliente para API de alertas
 * Según diagrama UML - AlertaService
 */
import axios from 'axios';

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || 'http://localhost:5000';

const alertaService = {
  /**
   * Obtiene todas las alertas con filtros opcionales
   * @param {Object} filters - Filtros: caso_id, camara_id, fecha_inicio, fecha_fin, estado, prioridad
   * @returns {Promise} - Lista de alertas
   */
  async getAlertas(filters = {}) {
    try {
      const params = new URLSearchParams();
      
      if (filters.caso_id) params.append('caso_id', filters.caso_id);
      if (filters.camara_id) params.append('camara_id', filters.camara_id);
      if (filters.fecha_inicio) params.append('fecha_inicio', filters.fecha_inicio);
      if (filters.fecha_fin) params.append('fecha_fin', filters.fecha_fin);
      if (filters.estado) params.append('estado', filters.estado);
      if (filters.prioridad) params.append('prioridad', filters.prioridad);

      const response = await axios.get(`${API_BASE_URL}/alertas?${params}`, {
        withCredentials: true
      });

      return response.data;
    } catch (error) {
      console.error('Error obteniendo alertas:', error);
      throw error;
    }
  },

  /**
   * Obtiene alertas en formato GeoJSON para mapas
   * Según UML: getAlertasPorFecha, getAlertasPorCamara, getAlertasPorCaso
   * @param {Object} filters - Filtros de búsqueda
   * @returns {Promise} - FeatureCollection GeoJSON
   */
  async getAlertasGeoJSON(filters = {}) {
    try {
      const params = new URLSearchParams();
      
      if (filters.caso_id) params.append('caso_id', filters.caso_id);
      if (filters.camara_id) params.append('camara_id', filters.camara_id);
      if (filters.fecha_inicio) params.append('fecha_inicio', filters.fecha_inicio);
      if (filters.fecha_fin) params.append('fecha_fin', filters.fecha_fin);

      const response = await axios.get(`${API_BASE_URL}/alertas/geojson?${params}`, {
        withCredentials: true
      });

      return response.data;
    } catch (error) {
      console.error('Error obteniendo GeoJSON de alertas:', error);
      throw error;
    }
  },

  /**
   * Obtiene línea temporal de movimientos de un caso
   * @param {number} caso_id - ID del caso
   * @param {string} fecha_inicio - Fecha inicio (opcional)
   * @param {string} fecha_fin - Fecha fin (opcional)
   * @returns {Promise} - Timeline de alertas
   */
  async getTimeline(caso_id, fecha_inicio = null, fecha_fin = null) {
    try {
      const params = new URLSearchParams();
      params.append('caso_id', caso_id);
      
      if (fecha_inicio) params.append('fecha_inicio', fecha_inicio);
      if (fecha_fin) params.append('fecha_fin', fecha_fin);

      const response = await axios.get(`${API_BASE_URL}/alertas/timeline?${params}`, {
        withCredentials: true
      });

      return response.data;
    } catch (error) {
      console.error('Error obteniendo timeline:', error);
      throw error;
    }
  },

  /**
   * Obtiene una alerta específica por ID
   * @param {number} alertaId - ID de la alerta
   * @returns {Promise} - Datos de la alerta
   */
  async getAlertaById(alertaId) {
    try {
      const response = await axios.get(`${API_BASE_URL}/alertas/${alertaId}`, {
        withCredentials: true
      });

      return response.data;
    } catch (error) {
      console.error('Error obteniendo alerta:', error);
      throw error;
    }
  },

  /**
   * Actualiza el estado de una alerta
   * @param {number} alertaId - ID de la alerta
   * @param {string} estado - Nuevo estado: "PENDIENTE" | "REVISADA" | "FALSO_POSITIVO"
   * @returns {Promise} - Resultado de la operación
   */
  async updateEstado(alertaId, estado) {
    try {
      const response = await axios.patch(
        `${API_BASE_URL}/alertas/${alertaId}/estado`,
        { estado },
        { withCredentials: true }
      );

      return response.data;
    } catch (error) {
      console.error('Error actualizando estado:', error);
      throw error;
    }
  },

  /**
   * Marca una alerta como revisada
   * @param {number} alertaId - ID de la alerta
   * @returns {Promise} - Resultado de la operación
   */
  async marcarRevisada(alertaId) {
    try {
      const response = await axios.post(
        `${API_BASE_URL}/alertas/${alertaId}/revisar`,
        {},
        { withCredentials: true }
      );

      return response.data;
    } catch (error) {
      console.error('Error marcando como revisada:', error);
      throw error;
    }
  },

  /**
   * Marca una alerta como falso positivo
   * @param {number} alertaId - ID de la alerta
   * @returns {Promise} - Resultado de la operación
   */
  async marcarFalsoPositivo(alertaId) {
    try {
      const response = await axios.post(
        `${API_BASE_URL}/alertas/${alertaId}/falso-positivo`,
        {},
        { withCredentials: true }
      );

      return response.data;
    } catch (error) {
      console.error('Error marcando como falso positivo:', error);
      throw error;
    }
  },

  /**
   * Obtiene estadísticas de alertas
   * @returns {Promise} - Estadísticas generales
   */
  async getEstadisticas() {
    try {
      const response = await axios.get(`${API_BASE_URL}/alertas/estadisticas`, {
        withCredentials: true
      });

      return response.data;
    } catch (error) {
      console.error('Error obteniendo estadísticas:', error);
      throw error;
    }
  },

  /**
   * Obtiene alertas pendientes de revisión
   * @returns {Promise} - Lista de alertas pendientes
   */
  async getAlertasPendientes() {
    try {
      const response = await axios.get(`${API_BASE_URL}/alertas/pendientes`, {
        withCredentials: true
      });

      return response.data;
    } catch (error) {
      console.error('Error obteniendo alertas pendientes:', error);
      throw error;
    }
  },

  /**
   * Obtiene alertas de alta prioridad
   * @returns {Promise} - Lista de alertas de alta prioridad
   */
  async getAlertasAltaPrioridad() {
    try {
      const response = await axios.get(`${API_BASE_URL}/alertas/alta-prioridad`, {
        withCredentials: true
      });

      return response.data;
    } catch (error) {
      console.error('Error obteniendo alertas de alta prioridad:', error);
      throw error;
    }
  }
};

export default alertaService;

File: facefind_back/services/alerta_service.py
"""
AlertaService - Servicio para gestión de alertas
Según diagrama UML: +crearAlerta(Date,float,float,float,ICamera,String,Frame,boolean):Alerta
                    +getAlertasPorFecha(Date,Date):List<Alerta>
                    +getAlertasPorCamara(ICamera):List<Alerta>
                    +getAlertasPorCaso(Caso):List<Alerta>
"""
from typing import Optional, List, Dict
from datetime import datetime
from models.alerta import Alerta
from models.frame import Frame
from models.enums import EstadoAlerta, PrioridadAlerta
from services.supabase_client import supabase


class AlertaService:
    """
    Servicio para crear y gestionar alertas del sistema
    Según diagrama UML
    """

    @staticmethod
    def crearAlerta(
        timestamp: datetime,
        confidence: float,
        latitud: float,
        longitud: float,
        camara_id: int,
        status: str,
        caso_id: int,
        frame: Frame,
        falso_positivo: bool = False
    ) -> Alerta:
        """
        Crea una nueva alerta en el sistema
        Según UML: +crearAlerta(timestamp:Date, confidence:float, latitud:float, 
                                longitud:float, camera:ICamera, status:String, 
                                imagenCaptura:Frame, falsoPositivo:boolean):Alerta

        Args:
            timestamp: Fecha y hora de la detección
            confidence: Nivel de confianza/similitud (0.0 - 1.0)
            latitud: Coordenada latitud de detección
            longitud: Coordenada longitud de detección
            camara_id: ID de la cámara (ICamera en UML)
            status: Estado de la alerta (String en UML)
            caso_id: ID del caso asociado
            frame: Frame capturado (imagenCaptura en UML)
            falso_positivo: Boolean indicando si es falso positivo

        Returns:
            Alerta creada
        """
        # Calcular prioridad basada en la similitud
        if confidence >= 0.85:
            prioridad = PrioridadAlerta.ALTA
        elif confidence >= 0.70:
            prioridad = PrioridadAlerta.MEDIA
        else:
            prioridad = PrioridadAlerta.BAJA

        # Convertir status string a EstadoAlerta
        estado = EstadoAlerta.from_string(status) if status else EstadoAlerta.PENDIENTE

        # Obtener ubicación de la cámara
        ubicacion = None
        try:
            camara_response = supabase.table("Camara").select("ubicacion").eq("id", camara_id).single().execute()
            if camara_response.data:
                ubicacion = camara_response.data.get("ubicacion")
        except Exception as e:
            print(f"⚠️ No se pudo obtener ubicación de cámara: {e}")

        # Crear objeto Alerta
        alerta = Alerta(
            caso_id=caso_id,
            camara_id=camara_id,
            timestamp=timestamp,
            similitud=confidence,
            prioridad=prioridad,
            estado=estado,
            ubicacion=ubicacion,
            latitud=latitud,
            longitud=longitud,
            imagen_captura=frame,
            imagen_bytes=frame.to_bytes() if frame else None,
            falso_positivo=falso_positivo
        )

        # Guardar en base de datos
        try:
            alerta_guardada = AlertaService._guardar_en_bd(alerta)
            return alerta_guardada
        except Exception as e:
            print(f"Error guardando alerta: {e}")
            raise

    @staticmethod
    def _guardar_en_bd(alerta: Alerta) -> Alerta:
        """
        Guarda la alerta en la base de datos

        Args:
            alerta: Objeto Alerta a guardar

        Returns:
            Alerta con ID asignado
        """
        data = {
            "caso_id": alerta.caso_id,
            "camara_id": alerta.camara_id,
            "timestamp": alerta.timestamp.isoformat() if isinstance(alerta.timestamp, datetime) else alerta.timestamp,
            "similitud": alerta.similitud,
            "ubicacion": alerta.ubicacion,
            "latitud": alerta.latitud,
            "longitud": alerta.longitud,
            "estado": alerta.estado.to_string(),
            "prioridad": alerta.prioridad.to_string(),
            "imagen": alerta.imagen_bytes,
            "falso_positivo": alerta.falso_positivo
        }

        # Agregar horarios si existen
        if alerta.horario_inicio:
            data["horario_inicio"] = alerta.horario_inicio.isoformat() if isinstance(alerta.horario_inicio, datetime) else alerta.horario_inicio
        if alerta.horario_fin:
            data["horario_fin"] = alerta.horario_fin.isoformat() if isinstance(alerta.horario_fin, datetime) else alerta.horario_fin

        response = supabase.table("Alerta").insert(data).execute()

        if not response.data:
            raise Exception("Error guardando alerta en BD")

        # Crear nueva instancia con el ID asignado
        alerta_data = response.data[0]
        return Alerta.from_dict(alerta_data)

    @staticmethod
    def obtener_alerta_por_id(alerta_id: int) -> Optional[Alerta]:
        """
        Obtiene una alerta por su ID

        Args:
            alerta_id: ID de la alerta

        Returns:
            Alerta o None si no existe
        """
        try:
            response = supabase.table("Alerta").select("*").eq("id", alerta_id).single().execute()

            if not response.data:
                return None

            return Alerta.from_dict(response.data)
        except Exception as e:
            print(f"Error obteniendo alerta: {e}")
            return None

    @staticmethod
    def obtener_alertas_por_caso(caso_id: int) -> List[Alerta]:
        """
        Obtiene todas las alertas de un caso

        Args:
            caso_id: ID del caso

        Returns:
            Lista de alertas
        """
        try:
            response = supabase.table("Alerta").select("*").eq("caso_id", caso_id).order("timestamp", desc=True).execute()

            if not response.data:
                return []

            return [Alerta.from_dict(data) for data in response.data]
        except Exception as e:
            print(f"Error obteniendo alertas del caso: {e}")
            return []

    @staticmethod
    def obtener_alertas_pendientes() -> List[Alerta]:
        """
        Obtiene todas las alertas pendientes de revisión

        Returns:
            Lista de alertas pendientes
        """
        try:
            response = supabase.table("Alerta").select("*").eq("estado", "PENDIENTE").order("timestamp", desc=True).execute()

            if not response.data:
                return []

            return [Alerta.from_dict(data) for data in response.data]
        except Exception as e:
            print(f"Error obteniendo alertas pendientes: {e}")
            return []

    @staticmethod
    def obtener_alertas_alta_prioridad() -> List[Alerta]:
        """
        Obtiene todas las alertas de alta prioridad

        Returns:
            Lista de alertas de alta prioridad
        """
        try:
            response = supabase.table("Alerta").select("*").eq("prioridad", "ALTA").order("timestamp", desc=True).execute()

            if not response.data:
                return []

            return [Alerta.from_dict(data) for data in response.data]
        except Exception as e:
            print(f"Error obteniendo alertas de alta prioridad: {e}")
            return []

    @staticmethod
    def actualizar_estado_alerta(alerta_id: int, nuevo_estado: EstadoAlerta) -> bool:
        """
        Actualiza el estado de una alerta

        Args:
            alerta_id: ID de la alerta
            nuevo_estado: Nuevo estado

        Returns:
            True si se actualizó correctamente
        """
        try:
            supabase.table("Alerta").update({
                "estado": nuevo_estado.to_string()
            }).eq("id", alerta_id).execute()
            return True
        except Exception as e:
            print(f"Error actualizando estado de alerta: {e}")
            return False

    @staticmethod
    def marcar_como_revisada(alerta_id: int) -> bool:
        """
        Marca una alerta como revisada

        Args:
            alerta_id: ID de la alerta

        Returns:
            True si se marcó correctamente
        """
        return AlertaService.actualizar_estado_alerta(alerta_id, EstadoAlerta.REVISADA)

    @staticmethod
    def marcar_como_falso_positivo(alerta_id: int) -> bool:
        """
        Marca una alerta como falso positivo

        Args:
            alerta_id: ID de la alerta

        Returns:
            True si se marcó correctamente
        """
        return AlertaService.actualizar_estado_alerta(alerta_id, EstadoAlerta.FALSO_POSITIVO)

    @staticmethod
    def obtener_estadisticas_alertas() -> Dict:
        """
        Obtiene estadísticas de alertas

        Returns:
            Diccionario con estadísticas
        """
        try:
            response = supabase.table("Alerta").select("estado, prioridad").execute()

            if not response.data:
                return {
                    "total": 0,
                    "por_estado": {},
                    "por_prioridad": {}
                }

            data = response.data
            stats = {
                "total": len(data),
                "por_estado": {},
                "por_prioridad": {}
            }

            for alerta in data:
                estado = alerta.get("estado", "PENDIENTE")
                prioridad = alerta.get("prioridad", "MEDIA")

                stats["por_estado"][estado] = stats["por_estado"].get(estado, 0) + 1
                stats["por_prioridad"][prioridad] = stats["por_prioridad"].get(prioridad, 0) + 1

            return stats
        except Exception as e:
            print(f"Error obteniendo estadísticas de alertas: {e}")
            return {"total": 0, "por_estado": {}, "por_prioridad": {}}

    @staticmethod
    def getAlertasPorFecha(fecha_inicio: datetime, fecha_fin: datetime) -> List[Alerta]:
        """
        Obtiene alertas filtradas por rango de fechas
        Según UML: +getAlertasPorFecha(inicio:Date, fin:Date):List<Alerta>

        Args:
            fecha_inicio: Fecha de inicio del rango
            fecha_fin: Fecha de fin del rango

        Returns:
            Lista de alertas en el rango especificado
        """
        try:
            # Query con JOIN para obtener datos de la cámara (ubicación, latitud, longitud)
            response = supabase.table("Alerta")\
                .select("*, Camara(id, ubicacion, latitud, longitud, ip, type)")\
                .gte("timestamp", fecha_inicio.isoformat())\
                .lte("timestamp", fecha_fin.isoformat())\
                .order("timestamp", desc=True)\
                .execute()

            if not response.data:
                return []

            # Enriquecer alertas con datos de cámara
            alertas = []
            for data in response.data:
                # Extraer datos de cámara del JOIN
                if "Camara" in data and data["Camara"]:
                    camara = data["Camara"]
                    data["latitud"] = camara.get("latitud") or data.get("latitud")
                    data["longitud"] = camara.get("longitud") or data.get("longitud")
                    if not data.get("ubicacion"):
                        data["ubicacion"] = camara.get("ubicacion")
                
                alertas.append(Alerta.from_dict(data))
            
            return alertas
        except Exception as e:
            print(f"Error obteniendo alertas por fecha: {e}")
            return []

    @staticmethod
    def getAlertasPorCamara(camara_id: int) -> List[Alerta]:
        """
        Obtiene todas las alertas de una cámara específica
        Según UML: +getAlertasPorCamara(camara:ICamera):List<Alerta>

        Args:
            camara_id: ID de la cámara (ICamera en UML)

        Returns:
            Lista de alertas de la cámara
        """
        try:
            # Query con JOIN para obtener datos de la cámara
            response = supabase.table("Alerta")\
                .select("*, Camara(id, ubicacion, latitud, longitud, ip, type)")\
                .eq("camara_id", camara_id)\
                .order("timestamp", desc=True)\
                .execute()

            if not response.data:
                return []

            # Enriquecer alertas con datos de cámara
            alertas = []
            for data in response.data:
                if "Camara" in data and data["Camara"]:
                    camara = data["Camara"]
                    data["latitud"] = camara.get("latitud") or data.get("latitud")
                    data["longitud"] = camara.get("longitud") or data.get("longitud")
                    if not data.get("ubicacion"):
                        data["ubicacion"] = camara.get("ubicacion")
                
                alertas.append(Alerta.from_dict(data))
            
            return alertas
        except Exception as e:
            print(f"Error obteniendo alertas de cámara: {e}")
            return []

    @staticmethod
    def getAlertasPorCaso(caso_id: int) -> List[Alerta]:
        """
        Obtiene todas las alertas de un caso (implementación OOP según UML)
        Según UML: +getAlertasPorCaso(caso:Caso):List<Alerta>

        Args:
            caso_id: ID del caso

        Returns:
            Lista de alertas del caso
        """
        try:
            # Query con JOIN para obtener datos de la cámara
            response = supabase.table("Alerta")\
                .select("*, Camara(id, ubicacion, latitud, longitud, ip, type)")\
                .eq("caso_id", caso_id)\
                .order("timestamp", desc=True)\
                .execute()

            if not response.data:
                return []

            # Enriquecer alertas con datos de cámara
            alertas = []
            for data in response.data:
                if "Camara" in data and data["Camara"]:
                    camara = data["Camara"]
                    data["latitud"] = camara.get("latitud") or data.get("latitud")
                    data["longitud"] = camara.get("longitud") or data.get("longitud")
                    if not data.get("ubicacion"):
                        data["ubicacion"] = camara.get("ubicacion")
                
                alertas.append(Alerta.from_dict(data))
            
            return alertas
        except Exception as e:
            print(f"Error obteniendo alertas del caso: {e}")
            return []

    @staticmethod
    def obtener_alertas_geojson(caso_id: Optional[int] = None, 
                                 fecha_inicio: Optional[datetime] = None,
                                 fecha_fin: Optional[datetime] = None,
                                 camara_id: Optional[int] = None) -> Dict:
        """
        Obtiene alertas en formato GeoJSON para visualización en mapas
        
        Args:
            caso_id: Filtro opcional por caso
            fecha_inicio: Filtro opcional por fecha inicio
            fecha_fin: Filtro opcional por fecha fin
            camara_id: Filtro opcional por cámara
            
        Returns:
            FeatureCollection GeoJSON con las alertas
        """
        try:
            # Construir query base con JOIN
            query = supabase.table("Alerta")\
                .select("*, Camara(id, ubicacion, latitud, longitud, ip, type), Caso(id, status, PersonaDesaparecida(nombre_completo))")

            # Aplicar filtros
            if caso_id:
                query = query.eq("caso_id", caso_id)
            if camara_id:
                query = query.eq("camara_id", camara_id)
            if fecha_inicio:
                query = query.gte("timestamp", fecha_inicio.isoformat())
            if fecha_fin:
                query = query.lte("timestamp", fecha_fin.isoformat())

            response = query.order("timestamp", desc=True).execute()

            if not response.data:
                return {
                    "type": "FeatureCollection",
                    "features": []
                }

            # Convertir a GeoJSON
            features = []
            for data in response.data:
                # Obtener coordenadas de la cámara
                lat = data.get("latitud")
                lon = data.get("longitud")
                
                if "Camara" in data and data["Camara"]:
                    camara = data["Camara"]
                    lat = camara.get("latitud") or lat
                    lon = camara.get("longitud") or lon

                # Solo incluir si tiene coordenadas válidas
                if lat is not None and lon is not None:
                    # Obtener nombre de persona desaparecida
                    persona_nombre = "Desconocido"
                    if "Caso" in data and data["Caso"]:
                        caso = data["Caso"]
                        if "PersonaDesaparecida" in caso and caso["PersonaDesaparecida"]:
                            persona_nombre = caso["PersonaDesaparecida"].get("nombre_completo", "Desconocido")

                    feature = {
                        "type": "Feature",
                        "geometry": {
                            "type": "Point",
                            "coordinates": [float(lon), float(lat)]  # GeoJSON usa [lon, lat]
                        },
                        "properties": {
                            "id": data.get("id"),
                            "caso_id": data.get("caso_id"),
                            "camara_id": data.get("camara_id"),
                            "timestamp": data.get("timestamp"),
                            "similitud": float(data.get("similitud", 0)),
                            "confidence": float(data.get("similitud", 0)),
                            "estado": data.get("estado"),
                            "prioridad": data.get("prioridad"),
                            "ubicacion": data.get("ubicacion"),
                            "falso_positivo": data.get("falso_positivo", False),
                            "persona_nombre": persona_nombre
                        }
                    }
                    features.append(feature)

            return {
                "type": "FeatureCollection",
                "features": features
            }

        except Exception as e:
            print(f"Error generando GeoJSON de alertas: {e}")
            import traceback
            traceback.print_exc()
            return {
                "type": "FeatureCollection",
                "features": []
            }
